lib/connectors/mongodb/connector.rb

```diff
@@ -56,30 +70,166 @@ def initialize(configuration: {})
+      end
+
+      def check_find_and_aggregate_present!(config)
+        find_and_aggregate_present = find_present?(config) && aggregate_present?(config)
+        both_present_msg = '\'find\' and \'aggregate\' functions cannot be used at the same time for MongoDB. Please drop one from the configuration.'
+
+        raise Utility::InvalidFilterConfigError.new(both_present_msg) if find_and_aggregate_present
+      end
+
+      def call_db_function_on_collection(function, args, &on_doc_serialization)
         with_client do |client|
           # We do paging using skip().limit() here to make Ruby recycle the memory for each page pulled from the server after it's not needed any more.
           # This gives us more control on the usage of the memory (we can adjust PAGE_SIZE constant for that to decrease max memory consumption).
           # It's done due to the fact that usage of .find.each leads to memory leaks or overuse of memory - the whole result set seems to stay in memory
           # during the sync. Sometimes (not 100% sure) it even leads to a real leak, when the memory for these objects is never recycled.
-          cursor = client[@collection].find
+          cursor = client[@collection].send(function, *args)
           skip = 0
 
+          options = args[1]
+          found_overall = 0
+
+          # if no overall limit is specified by filtering use -1 to not break ingestion, when no overall limit is specified (found_overall is only increased,
+          # thus can never reach -1)
+          overall_limit = -1
+
+          unless options.nil?
+            # there could be a skip parameter defined for filtering
+            skip = Utility::Commons.return_if_present(options[:skip], 0)
+            # there could be a limit parameter defined for filtering -> used for an overall limit (not a page limit, which was introduced for memory optimization)
+            overall_limit = Utility::Commons.return_if_present(options[:limit], -1)
+          end
+
+          overall_limit_reached = false
+
           loop do
-            found_count = 0
+            found_in_page = 0
+
             view = cursor.skip(skip).limit(PAGE_SIZE)
             view.each do |document|
-              yield serialize(document)
-              found_count += 1
+              yield serialize(document) { on_doc_serialization.call if block_given? }
+
+              found_in_page += 1
+              found_overall += 1
+
+              overall_limit_reached = found_overall == overall_limit
+
+              break if overall_limit_reached
             end
 
-            break if found_count == 0
+            page_was_empty = found_in_page == 0
+
+            break if page_was_empty || overall_limit_reached
+
             skip += PAGE_SIZE
           end
         end
       end
 
-      private
+      def setup_aggregate(advanced_config)
+        aggregate = advanced_config[:aggregate]
+
+        pipeline = aggregate[:pipeline]
+        options = extract_options(aggregate)
```

Comment:
Can `aggregate` have other options than `:pipeline`?

Reply:
Only `pipeline` and `options`: https://www.mongodb.com/docs/ruby-driver/current/api/Mongo/Collection.html#aggregate-instance_method